pkg/backend/identityapi/register_client.go

```diff
@@ -54,22 +68,106 @@ func NewIdentityRegisterClient(
 	}
+	icfg := identity.NewConfiguration()
+	icfg.BasePath = svcUrl + identityPath
+	icfg.Debug = true
+	// TODO: add the global HTTP client here
+	// icfg.HTTPClient = httpClient
+	identityClient := identity.NewAPIClient(icfg)
 	return &registerClient{
 		svcUrl:           strings.TrimSuffix(svcUrl, "/"),
 		licenseKey:       licenseKey,
 		userAgent:        userAgent,
 		httpClient:       httpClient,
 		compressionLevel: compressionLevel,
+		apiClient:        identityClient.DefaultApi,
 	}, nil
 }
 
+func (rc *registerClient) RegisterProtocolEntities(agentEntityID entity.ID, entities []protocol.Entity) (resp RegisterBatchEntityResponse, duration time.Duration, err error) {
+
+	ctx := context.Background()
+
+	registerRequests := make([]identity.RegisterRequest, len(entities))
+
+	for i := range entities {
+		registerRequest := identity.RegisterRequest{
+			EntityType:  entities[i].Type,
+			EntityName:  entities[i].Name,
+			DisplayName: entities[i].DisplayName,
+			Metadata:    convertMetadataToMapStringString(entities[i].Metadata),
+		}
+		registerRequests[i] = registerRequest
+	}
+
+	localVarOptionals := &identity.RegisterBatchPostOpts{
+		XNRIAgentEntityId: optional.NewInt64(int64(agentEntityID)),
+	}
+
+	apiReps, httpResp, err := rc.apiClient.RegisterBatchPost(ctx, rc.userAgent, rc.licenseKey, registerRequests, localVarOptionals)
+	if err != nil {
+		rlog.
+			WithError(err).
+			WithField("XNRIAgentEntityId", agentEntityID).
+			WithField("status", httpResp.StatusCode).
+			WithField("RegisterRequests", registerRequests).
+			Error("Something went wrong")
+		return resp, time.Second, err // TODO add right duration
+	}
+
+	resp = make(RegisterBatchEntityResponse, len(apiReps))
+
+	for i := range apiReps {
+		resp[i] = RegisterEntityResponse{
+			ID:   entity.ID(apiReps[i].EntityId),
+			Key:  entity.Key(apiReps[i].EntityName),
+			Name: apiReps[i].EntityName,
+		}
+	}
+
+	return resp, time.Second, err
+}
+
+func (rc *registerClient) RegisterEntity(agentEntityID entity.ID, ent protocol.Entity) (resp RegisterEntityResponse, err error) {
+
+	ctx := context.Background()
+	registerRequest := identity.RegisterRequest{
+		EntityType:  ent.Type,
+		EntityName:  ent.Name,
+		DisplayName: ent.DisplayName,
+		Metadata:    convertMetadataToMapStringString(ent.Metadata),
+	}
+	localVarOptionals := &identity.RegisterPostOpts{
+		XNRIAgentEntityId: optional.NewInt64(int64(agentEntityID)),
+	}
+
+	apiReps, httpResp, err := rc.apiClient.RegisterPost(ctx, rc.userAgent, rc.licenseKey, registerRequest, localVarOptionals)
+	if err != nil {
+		rlog.
+			WithError(err).
+			WithField("XNRIAgentEntityId", agentEntityID).
+			WithField("status", httpResp.StatusCode).
+			WithField("RegisterRequest", registerRequest).
+			Error("Something went wrong")
+		return resp, err
+	}
+
+	resp = RegisterEntityResponse{
+		ID:   entity.ID(apiReps.EntityId),
+		Key:  entity.Key(apiReps.EntityName),
+		Name: apiReps.EntityName,
+	}
```

Comment:
how about using a constructor?

Reply:
to be reused at https://github.com/newrelic/infrastructure-agent/pull/60/files#r465719330